core/services/relay/customendpoint/transmitter.go

```diff
@@ -0,0 +1,197 @@
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"math/big"
+	"net/url"
+
+	"github.com/pkg/errors"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promauto"
+	"github.com/shopspring/decimal"
+	"github.com/smartcontractkit/libocr/offchainreporting2/types"
+
+	"github.com/smartcontractkit/chainlink/core/bridges"
+	"github.com/smartcontractkit/chainlink/core/services/pipeline"
+	"github.com/smartcontractkit/chainlink/core/utils"
+)
+
+var _ types.ContractTransmitter = (*contractTracker)(nil)
+
+var (
+	promHTTPFetchTime = promauto.NewGaugeVec(prometheus.GaugeOpts{
+		Name: "custom_endpoint_http_fetch_time",
+		Help: "Time taken to fully execute the HTTP request",
+	},
+		[]string{"endpoint_target"},
+	)
+	promHTTPResponseBodySize = promauto.NewGaugeVec(prometheus.GaugeOpts{
+		Name: "custom_endpoint_http_response_body_size",
+		Help: "Size (in bytes) of the HTTP response body",
+	},
+		[]string{"endpoint_target"},
+	)
+)
+
+// Transmit uploads the result to customendpoint API endpoint, by making an HTTP call to
+// the customendpoint ExternalAdapter. The EA does the actual signing and uploading work.
+func (c *contractTracker) Transmit(
+	ctx context.Context,
+	reportCtx types.ReportContext,
+	report types.Report,
+	sigs []types.AttributedOnchainSignature,
+) error {
+
+	result, medianBigInt, err := c.getMedianFromReport(report)
+	if err != nil {
+		return err
+	}
+	c.transmittersWg.Add(1)
+
+	// Don't block the current thread for transmitting results to the endpoint
+	go c.doTransmit(ctx, result, medianBigInt, reportCtx.Epoch, reportCtx.Round)
+
+	return nil
+}
+
+// Returns the latest epoch from the last stored transmission.
+func (c *contractTracker) LatestConfigDigestAndEpoch(
+	ctx context.Context,
+) (
+	configDigest types.ConfigDigest,
+	epoch uint32,
+	err error,
+) {
+	digest, err := c.digester.configDigest()
+	c.ansLock.RLock()
+	defer c.ansLock.RUnlock()
+	return digest, c.storedAnswer.epoch, err
+}
+
+// TODO: Check if returning an item from StaticTransmitters value is good enough
+func (c *contractTracker) FromAccount() types.Account {
+	return StaticTransmitters[0]
+}
+
+func (c *contractTracker) doTransmit(
+	ctx context.Context,
+	result decimal.Decimal,
+	medianBigInt *big.Int,
+	epoch uint32,
+	round uint8) {
+	err := c.uploadResults(ctx, result)
+	if err != nil {
+		c.lggr.Warnw("Customendpoint Transmitter: Transmission failed",
+			"EndpointName: ", c.digester.EndpointName,
+			"EndpointTarget: ", c.digester.EndpointTarget,
+			"PayloadType: ", c.digester.PayloadType,
+			"Epoch: ", epoch,
+			"Round: ", round,
+			"Result: ", result.String(),
+			"Error:", err)
+	} else {
+		c.ansLock.RLock()
+		defer c.ansLock.RUnlock()
+
+		// Skip saving the storedAnswer if a more recent saved storedAnswer already exists
+		if epoch > c.storedAnswer.epoch || (epoch == c.storedAnswer.epoch && round > c.storedAnswer.round) {
+			c.ansLock.RLock()
+			defer c.ansLock.RUnlock()
```

Comment:
Is this meant to _read_ lock for the check, and _write_ lock for an update?

Reply:
Ahh, thanks again for pointing!
I wanted to do readLock for check, and writeLock for update.

Fixed now. All places which read, are using RLock/RUnlock.
This place, which is reading and updating, is now just doing a single Lock/Unlock.